File: SPL231-Assignment3-student-template-2/client/src/StompProtocol.cpp
#include <sstream>
#include <vector>
#include <map>
#include "../include/StompProtocol.h"
#include "../include/ConnectionHandler.h"
#include "../include/StompFrame.h"
#include "../include/event.h"

using namespace std;

StompProtocol::StompProtocol(ConnectionHandler& ch): 
mDisconnectRec(-1), mReceiptCounter(1), mSubId(0), user(), mConnectionHandler(&ch), commands(), subscriptions(), excpectedReciepts(), allReports()
{
    initCommands();
}

StompProtocol::StompProtocol(const StompProtocol& protocol): 
mDisconnectRec(-1), mReceiptCounter(1), mSubId(0), user(), mConnectionHandler(), commands(), subscriptions(), excpectedReciepts(), allReports()
{
    initCommands();
}

const StompProtocol& StompProtocol::operator=(const StompProtocol& protocol)
{
    initCommands();
    return *(this);
}

void StompProtocol::initCommands(){
    //Client
    commands["login"] = Command::login;
    //commands["connect"] = 1;
    commands["join"] = Command::join;
    commands["exit"] = Command::unjoin;
    commands["logout"] = Command::logout;
    commands["send"] = Command::sendMessage;
    commands["report"] = Command::report;
    commands["summary"] = Command::summary;

    //Server
    commands["MESSAGE"] = Frame::message;
    commands["RECEIPT"] = Frame::receipt;
    commands["CONNECTED"] = Frame::connected;
    commands["ERROR"] = Frame::error;
}

vector<string> StompProtocol::tokenize(string source, char delimiter){
    vector<string> tokens;
    string token;
    stringstream stream(source);

    while(getline(stream, token, delimiter)){
        tokens.push_back(token);
    }

    return tokens;
}

bool StompProtocol::validateCommand(vector<string> command){
    int unsigned expectedSize = 1;
    string structure;
    int type = -1;
    try{
        type = commands.at(command[0]);
    } catch(std::out_of_range& e){
        cout << "Invalid command type" << endl;
        return false;
    }
    
    switch (type) {

        case Command::login:
            expectedSize = 4;
            structure = "login {host:port} {username} {password}";
            break;

        case Command::join:
            expectedSize = 2;
            structure = "join {game_name}";
            if(subscriptions.count(command[1]) > 0){
                cout << "Client is already subscribed to " << command[1]  << " with id " << subscriptions[command[1]] << endl;
                return false;
            }
            break;
        case Command::unjoin:
            expectedSize = 2;
            structure = "exit {game_name}";
            try{
                subscriptions.at(command[1]);
            } catch(std::out_of_range& e){
                cout << "Invalid command. Client is not subscribed to " << command[1] << endl;
                return false;
            }
            break;

        case Command::sendMessage:
            expectedSize = 3;
            structure = "send {destination} {message}";
            if(command.size() < expectedSize){
                cout << "Invalid arguments. Expected: " << structure << endl;
                return false;
            } else 
                return true;
            break;

        case Command::report:
            expectedSize = 2;
            structure = "report {file}";
            break;

        case Command::summary:
            expectedSize = 4;
            structure = "summary {game_name} {user} {file}";
            break;
    }

        if(command.size() != expectedSize){
            cout << "Invalid arguments. Expected: " << structure << endl;
            return false;
        }
    return true;
}

vector<string> StompProtocol::processKeyboard(string msg) {
    vector<string> out;
    string s;

    if(msg == "")
        return out;

    vector<string> tokens = tokenize(msg, ' ');

    if(!mConnectionHandler->isLoggedIn() && tokens[0] != "login"){
        std::cout << "login first" << std::endl;
        return out;
    }

    if(!validateCommand(tokens))
        return out;

    switch (commands.at(tokens[0])) {
        
        case Command::login:
            out.push_back(login(tokens));
            break;
        
        case Command::join:
            out.push_back("SUBSCRIBE\nreceipt:" + to_string(mReceiptCounter) + "\nid:" + to_string(mSubId) + "\ndestination:" + tokens[1] + "\n\n\0");
            excpectedReciepts[mReceiptCounter] = make_tuple(subscribe, mSubId, tokens[1]);
            mSubId++;
            mReceiptCounter++;
            break;
        
        case Command::unjoin:
            out.push_back("UNSUBSCRIBE\nreceipt:" + to_string(mReceiptCounter) + "\nid:" + to_string(subscriptions.at(tokens[1])) + "\n\n\0");
            excpectedReciepts[mReceiptCounter] = make_tuple(unsubscribe, subscriptions.at(tokens[1]), tokens[1]);
            mReceiptCounter++;
            break;
        
        case Command::logout:
            excpectedReciepts[mReceiptCounter] = make_tuple(disconnect, 0, "");
            out.push_back(logout());
            break;
        
        case Command::sendMessage:
            s = tokens[2];
            for (int unsigned i=3; i<tokens.size(); i++)
                s += " " + tokens[i];
            out.push_back(send(tokens[1], s));
            break;

        case Command::report:
            out = report(tokens[1]);
            break;

        //Just a debug for now
        //TODO
        case Command::summary:
            for (auto const &pair1: allReports){
                for (auto const &pair2: allReports[make_tuple(get<0>(pair1.first),  get<1>(pair1.first))]){
                    string k = "(" + get<0>(pair1.first) + ", " + get<1>(pair1.first) + ")";
                    std::cout << k << " : (" << pair2.first << " : event )" << std::endl;
                }
            }
            break;

        default:
            std::cout << "Invalid frame code" << std::endl;
            break;
    }

    /* if(out != ""){
        std::cout << "=== SENT ===" << std::endl;
        std::cout << out << std::endl;
    } */
    return out;
}

void StompProtocol::processFrame(StompFrame newFrame) {
    //std::cout << "=== RECEIVED ===" << std::endl;
    //newFrame.printFrame(1);
    int recId;
    Event* event;
    string dest, user;
    switch(commands[newFrame.getCommand()]){

        case Frame::message:
            //newFrame.printFrame(1);
            //std::cout << std::endl << newFrame.getBody() << std::endl;
            event = new Event(newFrame.getBody());
            dest = newFrame.getHeaderValue("destination");
            user = extractUser(newFrame.getBody());

            std::cout << std::endl << "destination: " << dest << std::endl << "user: " << user << std::endl << std::endl;
            event->printEvent();
            //TODO: insert new event to allReports
            allReports[make_tuple(dest, user)][event->get_time()] = event;
        break;

        case Frame::receipt:
        recId = std::stoi(newFrame.getHeaderValue("receipt-id"));

        if (excpectedReciepts.count(recId)){
            Type t = get<0>(excpectedReciepts[recId]);
            int id = get<1>(excpectedReciepts[recId]);
            string d = get<2>(excpectedReciepts[recId]);
            switch(t){

                case disconnect:
                    mConnectionHandler->disconnect();
                    subscriptions.clear();
                    std::cout << "Disconnected" << std::endl;
                    break;
                
                case subscribe:
                    subscriptions[d] = id;
                    std::cout << "Subscribed to " + d << std::endl;
                    break;
                
                case unsubscribe:
                    subscriptions.erase(d);
                    std::cout << "Unsubscribed from " + d << std::endl;
                    break;

                default:
                    std::cout << "Unknown receipt received" + d << std::endl;
                    break;
            }

            excpectedReciepts.erase(recId);
        }
        break;

        case Frame::connected:
            mConnectionHandler->setLoggedIn(1);
            std::cout << "Login successful" << std::endl;
        break;

        //Error
        default:
            newFrame.printFrame(1);
        break;

    }
}

string StompProtocol::extractUser(const string& frameBody){
    string body(frameBody);
    int idx;

    idx = body.find("user:") + 6;
    body = body.substr(idx);
    idx = body.find('\n');
    
    return body.substr(0, idx);
}

string StompProtocol::login(vector<string> msg) {
    if(!mConnectionHandler -> isLoggedIn()) {
        string host = msg[1].substr(0, msg[1].find(":"));
        mConnectionHandler -> setHost(host);
        short port = std::stoi(msg[1].substr(msg[1].find(":") + 1, msg[1].length()));
        mConnectionHandler -> setPort(port);
        if (!mConnectionHandler -> connect()) {
            std::cerr << "Unable to connect " << host << ":" << port << std::endl;
            return "";
        } else {
            mConnectionHandler -> setName(msg[2]);
            string out = "CONNECT";
            out = out + "\n" + "accept-version:1.2" + "\n" + "host:stomp.cs.bgu.ac.il" + "\n" + "login:" + msg[2] + "\n" +
                "passcode:" + msg[3] + "\n" + "\n";
            return out;
        }
    }
    std::cout << "The client is already logged in, log out before trying again" << std::endl;
    return "";
}

string StompProtocol::logout() {
    string out = "DISCONNECT\nreceipt:" + to_string(mReceiptCounter) + "\n\n\0";
    mDisconnectRec = mReceiptCounter;
    mReceiptCounter++;
    return out;
}

string StompProtocol::send(const string& destination, const string& body){
    return "SEND\ndestination:" + destination + "\n\n" + body + "\n\0";
}

vector<string> StompProtocol::report(const string& file){
    vector<string> out;
    names_and_events allEvents = parseEventsFile(file);

    string dest = '/' + allEvents.team_a_name + '_' + allEvents.team_b_name;

    string head = "";
    head += "user: " + mConnectionHandler->getName() + '\n';
    head += "team a: " + allEvents.team_a_name + '\n';
    head += "team b: " + allEvents.team_b_name + '\n';

    for (Event event : allEvents.events){
        string body = "";
        body += "event name: " + event.get_name() + '\n';
        body += "time: " + to_string(event.get_time()) + '\n';

        body += "general game updates:\n";
        for(auto update : event.get_game_updates()){
            body += update.first + ": " + update.second + '\n';
        }

        body += "team a updates:\n";
        for(auto update : event.get_team_a_updates()){
            body += update.first + ": " + update.second + '\n';
        }

        body += "team b updates:\n";
        for(auto update : event.get_team_b_updates()){
            body += update.first + ": " + update.second + '\n';
        }

        body += "description:\n";
        body += event.get_discription();

        out.push_back(send(dest, head + body));
    }
    return out;
}

File: SPL231-Assignment3-student-template-2/server/src/main/java/bgu/spl/net/impl/stomp/StompProtocol.java
package bgu.spl.net.impl.stomp;

import bgu.spl.net.api.StompMessagingProtocol;
import bgu.spl.net.srv.ConnectionHandler;
import bgu.spl.net.srv.Connections;

import java.util.Hashtable;

public class StompProtocol implements StompMessagingProtocol<StompFrame> {
    private ConnectionsImpl<StompFrame> connections;
    private static final Hashtable<String, Integer> commandToInt = new Hashtable<>();

    private boolean shouldTerminate;
    private static int msgId;
    private static int nextConnectionId;

    public StompProtocol(){
        connections = ConnectionsImpl.getInstance();
        shouldTerminate = false;
        msgId = -1;
        nextConnectionId = 0;
        commandToInt.put("CONNECT", 1);
        commandToInt.put("SEND", 2);
        commandToInt.put("SUBSCRIBE", 3);
        commandToInt.put("UNSUBSCRIBE", 4);
        commandToInt.put("DISCONNECT", 5);
    }

    @Override
    public void start(ConnectionHandler<StompFrame> ch, int connectionId) {

    }

    private int assignMsgId(){
        msgId++;
        return msgId;
    }

    @Override
    public StompFrame process(StompFrame newFrame, ConnectionHandler<StompFrame> ch) {
        int connectionId = connections.connect(ch);

        System.out.println("CH with id: " + connectionId);
        System.out.println("=== Received ===");
        System.out.println(newFrame);

        switch (commandToInt.get(newFrame.getCommand())) {
            case 1 -> {
                //CONNECT
                connect(connectionId, newFrame);
            }

            case 2 -> {
                //SEND
                send(connectionId, newFrame);
            }

            case 3 -> {
                //SUBSCRIBE
                subscribe(connectionId, newFrame);
            }

            case 4 -> {
                //UNSUBSCRIBE
                unsubscribe(connectionId, newFrame);
            }

            case 5 -> {
                //DISCONNECT
                disconnect(connectionId, newFrame);
            }

            //Invalid command
            default -> {
                error(connectionId, "Invalid frame", newFrame.getCommand(), newFrame);
            }
        }
        return null;
    }

    private void connect(int connectionId, StompFrame frame){
        if (frame.getHeaders().containsKey("accept-version")) {
            if (frame.getHeaderValue("accept-version").equals("1.2")) {
                    if(connections.getUserNamePassword().containsKey(frame.getHeaderValue("login"))) {
                        if (connections.getUserNamePassword().get(frame.getHeaders().get("login")).equals(frame.getHeaders().get("passcode"))) {
                            Hashtable<String, String> sendHeaders = new Hashtable<>();
                            sendHeaders.put("version", frame.getHeaderValue("accept-version"));
                            connections.send(connectionId, new StompFrame("CONNECTED", sendHeaders, ""));
                            System.out.println(frame.getHeaderValue("login") + "connected");

                        } else {
                            error(connectionId, "Invalid login or password", "", frame);
                        }
                    }
                        else{
                            connections.register(frame.getHeaderValue("login"), frame.getHeaderValue("passcode"));
                            Hashtable<String, String> sendHeaders = new Hashtable<>();
                            sendHeaders.put("version", frame.getHeaderValue("accept-version"));
                            connections.send(connectionId, new StompFrame("CONNECTED", sendHeaders, ""));
                            System.out.println(frame.getHeaderValue("login") + " registered");
                        }
            } else {
                error(connectionId, "Invalid version", "", frame);
            }
        } else {
            error(connectionId, "Missing version", "", frame);
        }

    }

    private void send(int connectionId, StompFrame frame){
        if (frame.getHeaders().containsKey("destination")) {
            if (!frame.getBody().isEmpty()) {
                int subId = connections.getSubId(connectionId, frame.getHeaderValue("destination"));
                if(subId == -1){
                    error(connectionId, "Invalid Send", "You cannot sent message to an unsubscribed channel", frame);
                    return;
                }
                Hashtable<String, String> sendHeaders = new Hashtable<>();
                sendHeaders.put("destination", frame.getHeaderValue("destination"));
                sendHeaders.put("subscription", String.valueOf(subId));
                sendHeaders.put("message-id", String.valueOf(assignMsgId()));
                connections.send(frame.getHeaderValue("destination"), new StompFrame("MESSAGE", sendHeaders, frame.getBody()));
                if(frame.getHeaders().containsKey("receipt-id"))
                    receipt(connectionId, frame);
            } else {
                error(connectionId, "Missing body", "", frame);
            }
        } else {
            error(connectionId, "Missing destination", "", frame);
        }
    }

    private void subscribe(int connectionId, StompFrame frame){
        if (frame.getHeaders().containsKey("destination")) {
            if (frame.getHeaders().containsKey("id")) {
                if(!connections.subscribe(connectionId, Integer.parseInt(frame.getHeaderValue("id")), frame.getHeaderValue("destination")))
                    error(connectionId, "Subscription failed", "An error accured or you are already subscribed to this channel", frame);
                else if(frame.getHeaders().containsKey("receipt-id"))
                    receipt(connectionId, frame);

            } else {
                error(connectionId, "Missing id", "", frame);
            }
        } else {
            error(connectionId, "Missing destination", "", frame);
        }
    }

    private void unsubscribe(int connectionId, StompFrame frame){
        if (frame.getHeaders().containsKey("id")) {
            if(!connections.unsubscribe(connectionId, Integer.parseInt(frame.getHeaderValue("id"))))
                error(connectionId, "Invalid subscription id", "", frame);
            if(frame.getHeaders().containsKey("receipt-id"))
                receipt(connectionId, frame);
        } else {
            error(connectionId, "Missing id", "", frame);
        }
    }

    private void disconnect(int connectionId, StompFrame sourceFrame){
        if(sourceFrame.getHeaders().containsKey("receipt")){
            receipt(connectionId, sourceFrame);
            connections.disconnect(connectionId);
        } else {
            error(connectionId, "Missing receipt", "", sourceFrame);
        }
    }

    private void error(int connectionId, String message, String body, StompFrame sourceFrame){

        Hashtable<String, String> sendHeaders = new Hashtable<>();
        sendHeaders.put("message", message);
        if(sourceFrame.getHeaders().containsKey("receipt-id"))
            sendHeaders.put("receipt-id", message);
        if(sourceFrame.getHeaders().containsKey("message-id"))
            sendHeaders.put("message-id", message);
        connections.send(connectionId, new StompFrame("ERROR", sendHeaders, body));
        connections.disconnect(connectionId);
    }

    private void receipt(int connectionId, StompFrame sourceFrame){
        Hashtable<String, String> receiptHeaders = new Hashtable<>();
        receiptHeaders.put("receipt-id", sourceFrame.getHeaderValue("receipt"));
        connections.send(connectionId, new StompFrame("RECEIPT", receiptHeaders, ""));
    }

    @Override
    public void start(int connectionId, Connections<StompFrame> connections) {

    }

    @Override
    public boolean shouldTerminate() {
        return this.shouldTerminate;
    }
}
